<input type="number">: Normalize full-width input and enforce stricter validation

kyouhei-horizumi opened the pull request:
#### 6c84234e57fc7f1b0ccfa19f7e5d349e1a587684
<pre>
&lt;input type=&quot;number&quot;&gt;: Normalize full-width input and enforce stricter validation
<a href="https://bugs.webkit.org/show_bug.cgi?id=284363">https://bugs.webkit.org/show_bug.cgi?id=284363</a>

Reviewed by Aditya Keerthi.

This patch adds normalization support for full-width digits, the minus sign variants (U+FF0D, U+30FC, U+2212), and the full-width full stop (U+FF0E) by mapping them to their ASCII equivalents (e.g., U+002D for minus, U+002E for dot) when handling typed input in &lt;input type=number&gt; fields.
These characters are commonly produced by Japanese IMEs when entering numeric values.
In addition, like Chromium, disallowed characters are now rejected, and normalization is applied immediately after input.

Canonical link: <a href="https://commits.webkit.org/298613@main">https://commits.webkit.org/298613@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3a8b09a9db0a6a7d5809d4c3172c373c1e1f194c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116025 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35686 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26227 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122081 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66572 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117914 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36380 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44274 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88148 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118973 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29020 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104112 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68558 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28143 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22221 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65762 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/108134 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98411 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22357 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125231 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/114553 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42919 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32213 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96890 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43284 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100302 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96674 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41937 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19816 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18546 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42806 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48398 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/143237 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42272 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36926 "Found 7355 new JSC stress test failures: airjs-tests.yaml/stress-test.js.lockdown, basic-tests.yaml/stress-test.js.lockdown, cdjs-tests.yaml/main.js.lockdown, cdjs-tests.yaml/motion_test.js.lockdown, cdjs-tests.yaml/red_black_tree_test.js.lockdown, cdjs-tests.yaml/reduce_collision_set_test.js.lockdown, microbenchmarks/Float32Array-matrix-mult.js.lockdown, microbenchmarks/Int16Array-bubble-sort-with-byteLength.js.lockdown, microbenchmarks/Int16Array-bubble-sort.js.lockdown, microbenchmarks/Int16Array-load-int-mul.js.lockdown ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45607 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43977 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/c9d4536bc5225f494e990fae2c539fe15b738a81)<details>

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108865 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28525 "Build is in progress. Recent messages:") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114074 "Build is in progress. Recent messages:Running configuration") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110828 "Build is in progress. Recent messages:Running configuration") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/114074 "Build is in progress. Recent messages:Running configuration") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111813 "Build is in progress. Recent messages:Running configuration") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/114074 "Build is in progress. Recent messages:Running configuration") | 
| | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58775 "Build is in progress. Recent messages:") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/101403 "Build is in progress. Recent messages:Running configuration") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117194 "Build is in progress. Recent messages:Running configuration") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/107424 "Build is in progress. Recent messages:") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/117194 "Build is in progress. Recent messages:Running configuration") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36288 "Build is in progress. Recent messages:") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/117194 "Build is in progress. Recent messages:Running configuration") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-intel-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 playstation ](https://ews-build.webkit.org/#/builders/PlayStation-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35814 "Build is in progress. Recent messages:") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/131700 "Build is in progress. Recent messages:") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35515 "Build is in progress. Recent messages:") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/131700 "Build is in progress. Recent messages:") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38859 "Build is in progress. Recent messages:") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37201 "Build is in progress. Recent messages:") | | | </details>

annevk commented:
Thank you for improving this part of the web! If you fill out "New issue" at https://github.com/whatwg/html/issues/new/choose with the basic information I'm happy to copy the relevant people. I suspect people will agree though it might take a while to get everyone to comment, so you could do some of the other work in parallel.

kyouhei-horizumi commented:
I've opened a WHATWG HTML issue to propose this behavior formally:
https://github.com/whatwg/html/issues/11395

Thank you again for your feedback and support!

kyouhei-horizumi commented:
I've discovered a critical localization issue with the approach in this patch (and also in Chromium):

The Chromium code strips characters like , (comma) before the locale conversion step.
As a result, <input type="number" lang="fr-fr"> no longer converts "1,23" to "1.23" as expected.
This breaks locale-dependent numeric input for French and other locales that use comma as decimal separator.

Chromium 137 (before my full-width character changes) already had this regression, likely because there were no tests covering this scenario.

The current WebKit implementation does not have this issue, so the problem only appears after porting the Chromium logic.

I think it's better to pause this patch and re-evaluate the processing order:
1. Apply locale conversion first
2. Then strip any remaining invalid characters

I'll switch focus back to Chromium to investigate and address this upstream first.
Considering the broader impact on the web platform, fixing this in Chromium should take higher priority.

I will hold off on making any further changes in WebKit until that upstream work is complete.

Thanks—I was able to notice this because I ported the Chromium code and ran the existing WebKit tests. Without those tests, it would have been much harder to detect this regression.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e782662e706e5377ac9bc323393d11757d8dab68. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47095

lilyspiniolas commented:
The proposal makes sense to me; the displayed value should ideally match the value submitted, so if invalid values are not being submitted, I don't think they should display to the user.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3568f234a3b8422eb1ec0eb3712fbe6f102664fe. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47095

kyouhei-horizumi commented:
All planned changes for `<input type="number">` input normalization have now been completed.

# 📊 Summary of Improvements
This update adds full support for normalization of characters commonly produced by Japanese IMEs:

Full-width digits: `０１２３４５６７８９` → `0-9`

Minus sign variants:

`－` (U+FF0D)

`−` (U+2212)

`ー` (U+30FC)
→ All normalized to ASCII `-`

Full-width full stop: `．` (U+FF0E) → `.`

These characters are now automatically normalized at input time (keyboard and paste) for <input type="number"> fields.
(Direct assignment via .value remains untouched and must not accept non-ASCII characters as valid input.)

## 🔧 Additional Enhancements (aligned with Chromium behavior)
To improve consistency with the HTML specification and reduce user confusion, we revised WebKit’s input sanitization logic based on Chromium’s HandleBeforeTextInsertedEvent. Specifically, Chromium:

+ Normalizes full-width numerals, minus signs, and full stops to ASCII equivalents
+ Removes any characters not in the allowed set [0-9 . e E - +]
+ Validates character placement (e.g., only one decimal point, only one exponent character, signs in valid positions)

WebKit now incorporates equivalent logic, with variable names and function structure intentionally aligned with Chromium. This design choice was made to reduce the maintenance burden if we need to track upstream changes in the future.

On top of that, WebKit introduces stricter validation and refinements:

### 🌐 Locale-specific behavior is preserved

WebKit continues to support localized number formats such as using a comma (`,`) for decimal points in fr-FR. This has been verified via layout tests to ensure such input is handled correctly and not rejected.

### 🧠 Proactive rejection of ambiguous sequences

WebKit blocks inputs that Chromium currently allows but which result in `.value === ""`, creating misleading UI states.
For example:

+ In Chrome 138, `+1e-1` is visually accepted but invalid.
+ `-1-1` is also accepted despite having incorrect placement of signs.
WebKit now prevents these from being entered altogether.

#### 💡 At the same time, we do allow intermediate states that are expected during valid input composition:

For example, typing `-` or `e` results in a temporary empty .value, and this is permitted to support natural typing behavior.

#### ⌫ We deliberately do not block deletion-based transformations:

For instance, a user may delete parts of `-1e-1` to form `--1`; although .value becomes empty, this state is still allowed to appear temporarily in the UI, just as it is in Chromium.

## ⚠️ Breaking Changes

As a result of the stricter sanitization:

Clearly invalid characters (e.g., `a`) that were previously accepted and visually inserted into `<input type="number">` fields are now rejected and no longer appear.
(This matches Chromium's behavior.)

For example, pasting the string `1abc1` will now result in only `11` being inserted. Non-numeric characters are silently discarded.
(Again, consistent with Chromium.)

### Before (Previous WebKit behavior):
```html
<input type="number" id="test">
<script>
// User types or pastes "1abc1"
document.getElementById('test').value; // "" (invalid characters remained)
document.getElementById('test').validity.badInput; // true
</script>
```

### After (New WebKit behavior):
```html
<input type="number" id="test">
<script>
// User types or pastes "1abc1"
document.getElementById('test').value; // "11" (invalid characters silently removed)
document.getElementById('test').validity.badInput; // false
</script>
```

This stricter handling required changes to multiple layout tests that previously assumed such invalid characters would remain in the input field.

annevk commented:
It looks like you need a slight rebase and you want to replace `Reviewed by Alexey Proskuryakov.` with `Reviewed by NOBODY (OOPS!).` in the commit message. (The tools will take care of eventually replacing that with those that reviewed your patch.)

It's also WebKit style to not use `const` for local variables generally and to use `auto` where possible (at least when it's shorter than the name of the type and the type is not a smart pointer).

@lilyspiniolas and @pxlcoder can hopefully make time to review this.

lilyspiniolas commented on Source/WebCore/html/NumberInputType.cpp:
```diff
@@ -141,6 +143,110 @@ bool NumberInputType::typeMismatch() const
+    // Step 1: Check if normalization is needed
+    bool needsFiltering = false;
+    for (size_t i = 0; i < len; ++i) {
+        UChar ch = input[i];
```
```suggestion
        UChar char = input[i];
```

lilyspiniolas commented on Source/WebCore/html/NumberInputType.cpp:
```diff
@@ -141,6 +143,110 @@ bool NumberInputType::typeMismatch() const
+    // Step 3: Normalize and return a new String
+    StringBuilder builder;
+    builder.reserveCapacity(len);
+    for (size_t i = 0; i < len; ++i) {
```
(same comment regarding `unsigned`)

lilyspiniolas commented on Source/WebCore/html/NumberInputType.cpp:
```diff
@@ -141,6 +143,110 @@ bool NumberInputType::typeMismatch() const
+    StringBuilder builder;
+    builder.reserveCapacity(len);
+    for (size_t i = 0; i < len; ++i) {
+        UChar ch = input[i];
```
```suggestion
        const auto character = input[i];
```

lilyspiniolas commented on Source/WebCore/html/NumberInputType.cpp:
```diff
@@ -141,6 +143,110 @@ bool NumberInputType::typeMismatch() const
+
+ValueOrReference<String> NumberInputType::normalizeFullWidthNumberChars(const String& input) const
+{
+    const size_t len = input.length();
```
```suggestion
    const auto length = input.length();
```

lilyspiniolas commented on Source/WebCore/html/NumberInputType.cpp:
```diff
@@ -245,6 +351,189 @@ static bool isE(UChar ch)
+            } else if (!leftHalf.isEmpty()) {
+                if (hasE) {
+                    // Must be just after 'e'
+                    if (!isE(leftHalf[leftHalf.length() - 1]))
```
nit:
```suggestion
                    const auto lastCharacterOnLeftHalf = leftHalf[leftHalf.length() - 1];
                    if (!isE(lastCharacterOnLeftHalf))
```

lilyspiniolas reviewed (commented):
Looks mostly good to me, I would suggest going through everything again and updating types to `auto` wherever possible, and also using fullwords as well (i.e `length` instead of `len`, `character` instead of `ch`.)

lilyspiniolas commented:
The updated behavior of not displaying invalid character within number fields should be tested by executing the insert text command and then checking the inner text of the `contenteditable` div inside the shadow tree. You can use `internals.shadowRoot(numInputHere)` to get the shadow root and traverse the shadow tree as needed to check the inner div. `fast/forms/input-file-label.html` has examples of this

pxlcoder commented on Source/WebCore/html/NumberInputType.cpp:
```diff
@@ -141,6 +143,110 @@ bool NumberInputType::typeMismatch() const
+        return input;
+    }
+
+
```
Remove extra newline.

pxlcoder commented on Source/WTF/wtf/unicode/CharacterNames.h:
```diff
@@ -35,6 +35,7 @@ namespace WTF::Unicode {
 // more convenient for WebCore code that mostly uses UTF-16.
 
 constexpr char16_t HiraganaLetterSmallA = 0x3041;
+constexpr char16_t katakanaHiraganaProlongedSoundMark = 0x30FC;
```
This should be alphabetically sorted like the other constants.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9d6dba48d2b715369ed6860f4e831952cbcbed15. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47095

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b7e52083b322fabc0830d4689586e3d09cbc9e4b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47095

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a0fac0a7b5a44e49aab41b51069b7135dd9b3b5b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47095

webkit-early-warning-system commented:
EWS run on previous version of this PR (hash https://github.com/WebKit/WebKit/commit/aa2e6a8c95a8a9b55bc1e5a0a8ecca9bbcd9f9c7)<details>

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [⏳ 🧪 style ](https://ews-build.webkit.org/#/builders/Style-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31511 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Skipping applying patch since patch_id isn't provided; Running checkout-pull-request") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21988 "Failed to checkout and rebase branch from PR 47095") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117867 "Failed to checkout and rebase branch from PR 47095") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/117867 "Failed to checkout and rebase branch from PR 47095") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/138/builds/21988 "Failed to checkout and rebase branch from PR 47095") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/117867 "Failed to checkout and rebase branch from PR 47095") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/138/builds/21988 "Failed to checkout and rebase branch from PR 47095") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61707 "Failed to checkout and rebase branch from PR 47095") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/138/builds/21988 "Failed to checkout and rebase branch from PR 47095") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121140 "Build is in progress. Recent messages:Printed configuration") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/110407 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/121140 "Build is in progress. Recent messages:Printed configuration") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/138/builds/21988 "Failed to checkout and rebase branch from PR 47095") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/121140 "Build is in progress. Recent messages:Printed configuration") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/138/builds/21988 "Failed to checkout and rebase branch from PR 47095") | [⏳ 🛠 playstation ](https://ews-build.webkit.org/#/builders/PlayStation-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44271 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Running checkout-pull-request") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/41739 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | </details>

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2add54c5f89e9a909fce65773a8e5c2083b7fc99. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47095

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/07befcfc1b9e928cfd237e76ee2c82e5fd48f59a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47095

kyouhei-horizumi commented on Source/WebCore/html/NumberInputType.cpp:
```diff
@@ -141,6 +143,110 @@ bool NumberInputType::typeMismatch() const
+
+    // Step 1: Check if normalization is needed
+    bool needsFiltering = false;
+    for (size_t i = 0; i < len; ++i) {
```
Fixed.
https://github.com/WebKit/WebKit/pull/47095/commits/07befcfc1b9e928cfd237e76ee2c82e5fd48f59a#diff-9f23fede927b0353b10ccf6812e8f41f78df18240bedc5d79189889e394e92d1R154

kyouhei-horizumi commented on Source/WebCore/html/NumberInputType.cpp:
```diff
@@ -141,6 +143,110 @@ bool NumberInputType::typeMismatch() const
+
+    // Step 1: Check if normalization is needed
+    bool needsNormalization = false;
+    for (size_t i = 0; i < len; ++i) {
```
Fixed.
https://github.com/WebKit/WebKit/pull/47095/commits/07befcfc1b9e928cfd237e76ee2c82e5fd48f59a#diff-9f23fede927b0353b10ccf6812e8f41f78df18240bedc5d79189889e394e92d1R189

kyouhei-horizumi commented on Source/WebCore/html/NumberInputType.cpp:
```diff
@@ -141,6 +143,110 @@ bool NumberInputType::typeMismatch() const
+    if (!needsNormalization)
+        return input;
+
+    // Step 3: Normalize and return a new String
```
Fixed.
https://github.com/WebKit/WebKit/pull/47095/commits/07befcfc1b9e928cfd237e76ee2c82e5fd48f59a#diff-9f23fede927b0353b10ccf6812e8f41f78df18240bedc5d79189889e394e92d1R203

lilyspiniolas commented on LayoutTests/fast/forms/number/number-input-switch-from-text-clears-invalid-number-value.html:
```diff
@@ -0,0 +1,33 @@
+}
+</script>
+</body>
+</html>
```
```suggestion
</html>

```

lilyspiniolas commented:
Tests look good! I see that if an input with `type=text` that contains text has its type changed to `number`, the value of the input is correct, but I'm curious if the displayed value in the contenteditable div is correct as well?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/842695ed61713064f15ecc8b05957fb82fc0d672. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47095

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6229658fe9580964f4128033da701344d277e8a3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47095

lilyspiniolas commented:
Looks good to me! I don't have reviewer status so can only give this an unofficial r+ for now.

pxlcoder commented on Source/WebCore/html/NumberInputType.h:
```diff
@@ -47,7 +47,8 @@ class NumberInputType final : public TextFieldInputType {
     bool hasBadInput() const final;
-
+protected:
+    void handleBeforeTextInsertedEvent(BeforeTextInsertedEvent&) final;
```
```suggestion
    void handleBeforeTextInsertedEvent(BeforeTextInsertedEvent&) final;

```

pxlcoder replied:
But also, I think this can be moved into the `private` section?

kyouhei-horizumi replied:
Done.
https://github.com/WebKit/WebKit/compare/6229658fe9580964f4128033da701344d277e8a3..b25d140f014443a47c6deccecf55cee6dd361bc7
https://github.com/WebKit/WebKit/compare/b25d140f014443a47c6deccecf55cee6dd361bc7..54f31560bd5e3ac8664fda6eb8bf0108ffaadf5b

pxlcoder reviewed (commented):
I was going to also comment on the fact that the PR description contains a reviewer, but I noticed that the commit message does not, so that is fine.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b25d140f014443a47c6deccecf55cee6dd361bc7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47095

kyouhei-horizumi committed:
<input type="number">: Normalize full-width input and enforce stricter validation
https://bugs.webkit.org/show_bug.cgi?id=284363

Reviewed by Aditya Keerthi.

This patch adds normalization support for full-width digits, the minus sign variants (U+FF0D, U+30FC, U+2212), and the full-width full stop (U+FF0E) by mapping them to their ASCII equivalents (e.g., U+002D for minus, U+002E for dot) when handling typed input in <input type=number> fields.
These characters are commonly produced by Japanese IMEs when entering numeric values.
In addition, like Chromium, disallowed characters are now rejected, and normalization is applied immediately after input.

Canonical link: https://commits.webkit.org/298613@main

webkit-commit-queue commented:
Committed 298613@main (6c84234e57fc): <https://commits.webkit.org/298613@main>

Reviewed commits have been landed. Closing PR #47095 and removing active labels.

kyouhei-horizumi commented:
Hi @annevk, just wanted to say thanks!  
At first I thought I needed a second approval (confusing WebKit’s process with Chromium’s), and I was also stuck figuring out how to deal with the failing `jsc-armv7-tests` EWS run. In the end it turned out to be unrelated and the patch landed successfully.  
Really appreciate your guidance and the encouragement — it means a lot. Hope to contribute more going forward!
